stablehlo/reference/Api.cpp

```diff
@@ -137,6 +139,39 @@ LogicalResult validateEntrySignature(func::FuncOp func,
   return success();
 }
 
+// Specializes the shapes of arguments in function 'func' based on runtime input
+// values. If all argument types already have static shapes, this function does
+// nothing. Otherwise, it constructs a pipeline of MLIR passes to refine
+// argument shapes using the provided `inputs`.
+//
+// Args:
+//   module: The MLIR module containing the function.
+//   func: The function whose argument shapes need specialization.
+//   inputs: The runtime input values.
+//
+// Returns:
+//   A `LogicalResult` indicating success or failure of the shape
+//   refinement pipeline.
+LogicalResult removeDynamism(ModuleOp module, func::FuncOp func,
+                             ArrayRef<InterpreterValue> inputs) {
+  if (llvm::all_of(func.getArgumentTypes(), [](Type type) {
+        return llvm::cast<ShapedType>(type).hasStaticShape();
+      })) {
+    return success();
+  }
```

Comment:
Could there be dynamism in other parts of the program? If so, maybe we should still run the pipeline? Likely won't be too costly if it turns out that there is no refinement to be done.

Reply:
Thanks for bringing this. Yup that is possible.

```
func.func @main(%arg0: tensor<3x4xf32>, %arg1: tensor<3x4xf32>) -> tensor<?x?xf32> {
    %0 = stablehlo.add %arg0, %arg1 : (tensor<3x4xf32>, tensor<3x4xf32>) -> tensor<?x?xf32>
    return %0 : tensor<?x?xf32>
}
```

This currently will not be refined. 
The check for eligibility to call the pass-pipiline would be a iteration over the whole graph to check for any traces of dynamism.  That includes identifying ops like dynamic_*op with static operand/result type and const shape operand (This is not usual for someone to use such operations like that but theoretically possible). That would allow us to remove as many traces of dynamism as possible. 

Other  alternative is to call the pipeline regardless given the shape refinement pass already has the logic when to skip the refinement.

Thoughts?

Reply:
Right, checking for dynamism is kind of involved.

I think both approaches are fine, though I would tend to lean towards just calling the pipeline every time.

Reply:
done

Reply:
Note: Invoking the pass-pipeline unconditionally uncovers a few issues in the folding logic of shape refinement pass which can be addressed separately via  https://github.com/openxla/stablehlo/issues/2278